src/dmd/backend/inliner.d

```diff
@@ -0,0 +1,645 @@
+/**
+ * Function inliner.
+ *
+ * This is meant to replace the previous inliner, which inlined the front end AST.
+ * This inlines based on the intermediate code, after it is optimized,
+ * which is simpler and presumably can inline more functions.
+ * It does not yet have full functionality,
+ * - it inlines one expression only, statements that can be turned into ||, && or ?:
+ *   expressions are not done
+ * - it does not inline expressions with string literals in them, as these get turned into
+ *   local symbols which cannot be referenced from another object file
+ * - exception handling code for Win32 is not inlined
+ * - it does not give warnings for failed attempts at inlining pragma(inline, true) functions
+ * - it can only inline functions that have already been compiled
+ * - it cannot inline statements
+ *
+ * Compiler implementation of the
+ * $(LINK2 https://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (C) 2022 by The D Language Foundation, All Rights Reserved
+ *              Some parts based on an inliner from the Digital Mars C compiler.
+ * Authors:     $(LINK2 https://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 https://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/backend/inliner.d, backend/inliner.d)
+ */
+
+// C++ specific routines
+
+module dmd.backend.inliner;
```

Comment:
For the failing OSX test, looks like `core.thread.threadbase.Thread._slock` is used from two different DSOs.

```d
initLocks(0x10ae989e8)
initLocks(0x10ae98a48)
add(ThreadBase): 0x10ae989e8
Mutex.lock_nothrow(this:0x10ae989e8, 0x10ae98a00)
add(ThreadBase): 0x10ae989e8 locked
add(StackContext): 0x10ae989e8
Mutex.lock_nothrow(this:0x10ae989e8, 0x10ae98a00)
add(StackContext): 0x10ae989e8 locked
joinAll(): 0x10ae989e8
Mutex.lock_nothrow(this:0x10ae989e8, 0x10ae98a00)
joinAll(): 0x10ae989e8 locked
thread_suspendAll(): 0x10af76c58
Mutex.lock_nothrow(this:0x10af76c58, 0x10af76c70)
core.sync.exception.SyncError@(0): Unable to lock mutex.
```

This is not surprising, granted how DSOs work in OSX.  Unlike Linux where a BSS symbol defined in multiple DSOs is merged at run-time, on OSX it will resolve to a different address in each DSO.

```bash
user@darwin20:~/src/dlang/dmd/test$ nm test_results/dshell/dll/testdll | grep __D4core6thread10threadbase10ThreadBase6_slockG96v
00000001000399e8 B __D4core6thread10threadbase10ThreadBase6_slockG96v

user@darwin20:~/src/dlang/dmd/test$ nm test_results/dshell/dll/mydll.so | grep __D4core6thread10threadbase10ThreadBase6_slockG96v
0000000000057c58 B __D4core6thread10threadbase10ThreadBase6_slockG96v
```
The corrective thing to do would be to make these BSS globals weak.

I don't know why this doesn't fail on master, because this would be an issue with any shared library on OSX with dmd.

Reply:
@WalterBright the appropriate action here is just to disable the test, even [druntime](https://github.com/dlang/druntime/blob/master/src/rt/sections_osx_x86_64.d#L139) says that it shouldn't work (#14208).

Reply:
Thanks, @ibuclaw , you've just saved me hours of work! I appreciate it :-)